### PR TITLE
addpkg(x11/luanti{-common,-server}): moved from TUR 

### PR DESCRIPTION
**Pull request by robertkirkman**
Copied and pasted from https://github.com/termux-user-repository/tur/tree/master/tur/luanti

Previous documentation here:
- https://github.com/termux-user-repository/tur/issues/1295
- https://github.com/termux-user-repository/tur/pull/1296
- https://github.com/termux-user-repository/tur/pull/1331
- https://github.com/termux-user-repository/tur/pull/1327

Summary of key points:

- for better performance than normal `virglrenderer-android`, test using [termux-wsi-layer](https://github.com/termux/termux-packages/pull/22353) or [new virglrenderer](https://github.com/termux/termux-packages/pull/22385)

- to use touchscreen input, set "Touchscreen input mode: Direct touch" in Termux:X11 settings

image

- to use keyboard input, set "Prefer scancodes" in Termux:X11 settings

image

- to use mouse input, set "Capture external pointer devices" in Termux:X11 settings and then click once with the mouse

image

<details>
<summary>New recommended Termux:X11 settings for Luanti</summary>

```
$ termux-x11-preference list
"hideCutout"="false"
"displayStretch"="false"
"notificationButton1Action"="exit"
"forceOrientation"="reverse landscape"
"adjustResolution"="false"
"tapToMove"="false"
"enableAccessibilityServiceAutomatically"="false"
"backButtonAction"="toggle soft keyboard"
"xrMode"="false"
"adjustHeightForEK"="false"
"notificationTapAction"="open preferences"
"showAdditionalKbd"="false"
"stylusButtonContactModifierMode"="false"
"additionalKbdVisible"="true"
"opacityEKBar"="100"
"displayResolutionCustom"="1280x1024"
"stylusIsMouse"="false"
"clipboardEnable"="true"
"extra_keys_config"="[['ESC','/',{key: '-', popup: '|'},'HOME','UP','END','PGUP','PREFERENCES'], ['TAB','CTRL','ALT','LEFT','DOWN','RIGHT','PGDN','KEYBOARD']]"
"pauseKeyInterceptingWithEsc"="false"
"preferScancodes"="true"
"storeSecondaryDisplayPreferencesSeparately"="false"
"showIMEWhileExternalConnected"="true"
"useTermuxEKBarBehaviour"="false"
"showMouseHelper"="false"
"transformCapturedPointer"="No"
"displayResolutionExact"="1280x1024"
"keepScreenOn"="true"
"pointerCapture"="true"
"swipeDownAction"="toggle additional key bar"
"touchMode"="Direct touch"
"dexMetaKeyCapture"="false"
"scaleTouchpad"="true"
"hardwareKbdScancodesWorkaround"="true"
"notificationButton0Action"="open preferences"
"volumeUpAction"="no action"
"fullscreen"="true"
"enableGboardCJK"="false"
"Reseed"="false"
"PIP"="false"
"capturedPointerSpeedFactor"="100"
"displayResolutionMode"="native"
"swipeUpAction"="no action"
"filterOutWinkey"="false"
"displayScale"="100"
"volumeDownAction"="no action"
"showStylusClickOverride"="false"
```

</details>

- 6 patches are written entirely or mostly by me, but `0001-enable-egl-without-sdl2.patch` is not, it is this commit cherrypicked. https://github.com/minetest/minetest/commit/4838eb2f7de8477ef5be7e064d072954ff78e36c

> [!NOTE]
> In the future, Luanti might "transition back to SDL2" I think that a blocker for that in this build of luanti might be that building it with SDL2 prevented accessing EGL mode, at least for me. I have been building versions of SDL2 that might work with EGL when I test them, so that might come next eventually.

### CODE REVIEW

File: x11-packages/luanti/build.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+TERMUX_PKG_HOMEPAGE=https://www.luanti.org
+TERMUX_PKG_DESCRIPTION="An open source voxel game engine."
+TERMUX_PKG_LICENSE="LGPL-2.1"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=1:5.10.0
+TERMUX_PKG_REVISION=3
```
Comment (Biswa96) [inline x11-packages/luanti/build.sh:6]: As this is a new package in termux-packages repository, do we need revision and epoch in version?

Comment (robertkirkman) [reply x11-packages/luanti/build.sh:6]: I am not completely sure, licy183 helped me with that. @licy183 , do you know 
I assumed that, if anyone already has the package installed from TUR, but an older version, this is necessary to make sure they can still upgrade?

**Comment by Biswa96** [comment]
1. Some dependencies are used from bundled code. Should those be changed to system libraries?
2. Should the common files/assets be moved to a separate platform independent package? It would be shared with all architectures and reduce some package size. Those similar steps were done with supertuxkart in TUR.

File: x11-packages/luanti/build.sh (hunk 1 above)
Comment (truboxl) [reply x11-packages/luanti/build.sh:6]: As long as its one revision above TUR's then its fine

**Comment by robertkirkman** [comment]
1. not for Lua  because in my experience, Lua bundled inside games must identically match the same version of Lua used by other clients of the same game, down to minor revision numbers. For other dependencies it is probably ok to use a Termux package.
2. Yes AUR does that and has four packages, ["client" "server" "common" and "documentation"](https://aur.archlinux.org/packages?O=0&SeB=nd&K=luanti&outdated=&SB=p&SO=d&PP=50&submit=Go), but the whole package combined is only 6 MB, compared to the 655 MB of the supertuxkart packages combined which is very large. In luanti, the gamemodes and heavy assets are mostly downloaded dynamically through manual installation or the GUI. If you think it would still be best to separate the package into at least "client" "server" and "common", let me know and I will

File: x11-packages/luanti/build.sh (hunk 1 above)
Comment (robertkirkman) [reply x11-packages/luanti/build.sh]: Ok, it was the same revision as TUR, so now I changed it to be one revision higher.

**Comment by Biswa96** [comment]
> Which others did you notice?

According to previous build log https://github.com/termux/termux-packages/actions/runs/12031431167/job/33540963447#step:6:1054 those bundled dependencies were gmp, jsoncpp and luajit.

**Comment by robertkirkman** [comment]
I want the benchmarks, but they are not enabled by default, and neither is the server, I did not realize these things at first, so I will try creating subpackages like this:

- `luanti-client`
- `luanti-server`
- `luanti-common`
- `luanti-benchmarks`

and I'll update this with how that goes. I am not sure if the package is suited to splitting in termux-packages, but it makes sense to try it and check what happens.

File: x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+for platforms other than SurfaceFlinger. This prevents the error
+"CMake Error: failed to create symbolic link '/minetest': Permission denied"
+--- a/src/CMakeLists.txt
++++ b/src/CMakeLists.txt
+@@ -569,7 +569,7 @@ endif()
+ 
+ # When cross-compiling assume the user doesn't want to run the executable anyway,
+ # otherwise place it in <source dir>/bin/ since Luanti can only run from there.
+-if(NOT CMAKE_CROSSCOMPILING)
++if(NOT THIS_IS_FOR_SURFACEFLINGER_IN_OWOKITTYS_OPINION)
```
Comment (TomJo2000) [inline x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch:11]: That one patch line raises several questions.
All of them some variation of "what?"

- What are we actually patching here?
- What compelled you to choose that as the variable name?
- What does this have to do with SurfaceFlinger?

File: x11-packages/luanti/build.sh
Comment (TomJo2000) [inline x11-packages/luanti/build.sh:1]: So what's still in the main `luanti` package if the server and common files are split out as subpackages?

File: x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch (hunk 2 above)
Comment (robertkirkman) [reply x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch:11]: A better name for this change would be "DISABLING_THIS_BECAUSE_IT_IS_FOR_BUILDING_AN_APK" , and the idea of "SurfaceFlinger" originates from the fact that `.apk` files that launch ANativeWindow activities cannot be executed on devices where the program named `/system/bin/surfaceflinger` is not running in the background. 

On the other hand, unlike the preexisting build of Minetest that can be found on the F-Droid store, which is a `.apk` file, this build targets X11 and therefore it can hypothetically run connected to the X11 server from the package `tigervnc` inside termux-docker which is an environment where `/system/bin/surfaceflinger` is not present.

I will work on renaming this variable now.

File: x11-packages/luanti/build.sh
Comment (robertkirkman) [reply x11-packages/luanti/build.sh:1]: Good question, I will explain it like this, and if the packages need to be reorganized in a different way, this could be changed:

- the `luanti` package contains the `.desktop` shortcut, the icon, and a program `$PREFIX/bin/luanti` that actually has both a client and a server inside it. It can be made to behave as a server by launching it with the command `luanti --server`. The problem is just that this binary depends on a lot of packages for the graphical client as well, ` libjpeg-turbo`, `libxi` and others, so installing it requires installing those packages as runtime dependencies.
- The `luanti-server` package contains only the program `$PREFIX/bin/luantiserver`, which when launched, only implements the headless server functionality and does not have the graphical client built inside it. This means it does not depend on `libjpeg-turbo`, `libxi` or the other packages that are in the dependency list of the `luanti` package, but not the `luanti-server` package. This would be the package for use by someone who just wants to open Termux and run the Luanti server to connect to it remotely from Luanti clients on other devices, and doesn't have a desktop environment installed inside Termux.

- The `luanti-common` package contains platform-independent files that are necessary for both the client and the server.

File: x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch (hunk 2 above)
Comment (robertkirkman) [reply x11-packages/luanti/0002-fix-elf-directory-for-cross-compiling.patch]: @TomJo2000 to explain an answer to your question "What are we actually patching here?" in a way that is hopefully complete, precise and does not leave out any details:

Normally, Luanti does not set the `EXECUTABLE_OUTPUT_PATH` variable when cross-compiling is detected, because the person who wrote that section assumed/knew that they do not officially cross-compile for any platforms that require `EXECUTABLE_OUTPUT_PATH`.

Later on elsewhere in their `CMakeLists.txt`, they have this logic: "if not ANDROID... then if UNIX... then create a symlink in the `EXECUTABLE_OUTPUT_PATH`"

https://github.com/minetest/minetest/blob/6a1d22b2c516abbb9ce9670dedff47451317706f/src/CMakeLists.txt#L1045-L1054

The error that I saved in the comment of the patch, is the error that happens if `EXECUTABLE_OUTPUT_PATH` does not exist when that linked point in the `CMakeLists.txt` is reached.

**Comment by robertkirkman** [comment]
Brief suggestions for layer packages to use for the time being, on each of these Android 8+ devices, and the FPS you might expect to get in fullscreen with default settings on them; I listed the best-functioning layer package I have seen so far for Luanti + Termux:X11 on each of these devices, so if someone is trying to play and happens to have the same one as me and are using the newest possible Android ROM available for each of them, they should try the one listed first unless there have been more updates to them since I posted this.

- Galaxy S7 SM-G930U: `termux-wsi-layer`, ~20 fps
- Galaxy S9 SM-G960U: `termux-wsi-layer`, ~35 fps
- Galaxy S8+ SM-G955F: `virglrenderer-android`, ~15 fps
- Galaxy A70 SM-A705FN: `termux-wsi-layer`, ~35 fps
- Steam Deck OLED: `virglrenderer-android`, ~59 fps

**Comment by truboxl** [comment]
I think its better to put SUBPKG_BREAKS and SUBPKG_REPLACES fields with old version main package in the subpackages to handle split

**Comment by robertkirkman** [comment]
@truboxl I thought it might be a good idea as well, but I am not sure how to do it properly, maybe you have a good idea. In my test I tried this in `luanti-common.subpackage.sh`:

```
TERMUX_SUBPKG_INCLUDE="share/luanti/ share/man/ share/doc/"
TERMUX_SUBPKG_DESCRIPTION="Common files, including docs, shared between luanti and luanti-server"
TERMUX_SUBPKG_PLATFORM_INDEPENDENT=true
TERMUX_SUBPKG_BREAKS="luanti (<< 1:5.10.0-4)"
TERMUX_SUBPKG_REPLACES="luanti (<< 1:5.10.0-4)"
```

and also, separately and together, I tried this as well in `luanti-server.subpackage.sh`
```
TERMUX_SUBPKG_INCLUDE="bin/luantiserver bin/minetestserver"
TERMUX_SUBPKG_DESCRIPTION="Headless server for the Luanti voxel game engine."
TERMUX_SUBPKG_DEPENDS="jsoncpp, libandroid-spawn, libc++, libcurl, libgmp, libiconv, libpng, libsqlite, luajit, luanti-common, zstd"
TERMUX_SUBPKG_BREAKS="luanti (<< 1:5.10.0-4)"
TERMUX_SUBPKG_REPLACES="luanti (<< 1:5.10.0-4)"
```

However, unfortunately, if I build the packages with those changes, then I try to install them, this happens, which seems to prevent `luanti-server` from being installed without also installing the `luanti` package and pulling in all the same dependencies as the main package.

<details>
<summary>luanti-server depends luanti but it is not installable</summary>

```
~ $ pkg i ./luanti-common.deb 
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://mirrors.cbrx.io/apt/termux/termux-main: ok
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'luanti-common' instead of './luanti-common.deb'
The following package was automatically installed and is no longer required:
  xorg-xrandr
Use 'apt autoremove' to remove it.
The following packages will be REMOVED:
  luanti
The following NEW packages will be installed:
  luanti-common
0 upgraded, 1 newly installed, 1 to remove and 17 not upgraded.
Need to get 0 B/3344 kB of archives.
After this operation, 10.6 MB disk space will be freed.
Do you want to continue? [Y/n] ^C
~ $ pkg i ./luanti-server.deb 
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://mirrors.cbrx.io/apt/termux/termux-main: ok
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'luanti-server' instead of './luanti-server.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 luanti-server : Depends: luanti-common but it is not installable
                 Depends: luanti (= 1:5.10.0-4) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
~ $ pkg i ./luanti-server.deb ./luanti-common.deb 
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://mirrors.cbrx.io/apt/termux/termux-main: ok
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'luanti-server' instead of './luanti-server.deb'
Note, selecting 'luanti-common' instead of './luanti-common.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 luanti-server : Depends: luanti (= 1:5.10.0-4) but it is not installable
E: Unable to correct problems, you have held broken packages.
~ $ pkg i ./luanti-server.deb ./luanti-common.deb ./luanti.deb 
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://mirrors.cbrx.io/apt/termux/termux-main: ok
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'luanti-server' instead of './luanti-server.deb'
Note, selecting 'luanti-common' instead of './luanti-common.deb'
Note, selecting 'luanti' instead of './luanti.deb'
The following package was automatically installed and is no longer required:
  xorg-xrandr
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  luajit xdg-utils xorg-xprop
The following NEW packages will be installed:
  luajit luanti-common luanti-server xdg-utils xorg-xprop
The following packages will be upgraded:
  luanti
1 upgraded, 5 newly installed, 0 to remove and 17 not upgraded.
Need to get 324 kB/7374 kB of archives.
After this operation, 4370 kB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

</details>

Is there a way to do it without the main package becoming a dependency of the subpackages?

**Comment by truboxl** [comment]
Use TERMUX_SUBPKG_DEPEND_ON_PARENT=false if you think subpackage don't need the main package

**Comment by robertkirkman** [comment]
Thank you very much that works, I have added that now.

I would like to add a note that the reason why I have not added a `TERMUX_PKG_SERVICE_SCRIPT`  variable to the `luanti-server` package to make a fast launcher with a name like `luantid`, is because Luanti requires browsing, choosing, downloading and installing a gamemode folder into `~/.minetest/games` to launch the server, and prints a message with instructions for what to do if one is not present.

Arch Linux (AUR) does make the command `luantiserver --gameid devtest` work out of the box anyway by preinstalling the devtest gamemode with its `luanti-modern-common` package, but the reason I did not include that here is because if I did that, it would change the default title screen of the client to a checkerboard pattern with text "development test" and it's not immediately obvious that the white "+" at the bottom is for adding more gamemodes.

image

Instead of that (Arch Linux's default behavior), I thought the most natural and user-friendly default title screen for Termux users would be this one, so this is how I have it now - the button in the center leads to a menu where every gamemode, including devtest if needed, can be found and installed.

image

File: x11-packages/luanti/build.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+TERMUX_PKG_HOMEPAGE=https://www.luanti.org
+TERMUX_PKG_DESCRIPTION="An open source voxel game engine."
+TERMUX_PKG_LICENSE="LGPL-2.1"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=1:5.10.0
+TERMUX_PKG_REVISION=4
+TERMUX_PKG_SRCURL=https://github.com/minetest/minetest/archive/refs/tags/${TERMUX_PKG_VERSION:2}.zip
+TERMUX_PKG_SHA256=e74e994c0f1b188d60969477f553ad83b8ce20ee1e0e2dcd068120189cb0f56c
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_DEPENDS="freetype, jsoncpp, libandroid-spawn, libc++, libcurl, libglvnd-dev, libgmp, libjpeg-turbo, libiconv, libpng, libsqlite, libvorbis, libxi, luajit, luanti-common, openal-soft, xdg-utils, zstd"
```
Comment (truboxl) [inline x11-packages/luanti/build.sh:10]:
```suggestion
TERMUX_PKG_DEPENDS="freetype, jsoncpp, libandroid-spawn, libc++, libcurl, libgmp, libjpeg-turbo, libiconv, libluajit, libpng, libsqlite, libvorbis, libx11, libxi, luanti-common, openal-soft, opengl, xdg-utils, zlib, zstd"
```

File: x11-packages/luanti/luanti-server.subpackage.sh [hunk 4]
```diff
@@ -0,0 +1,6 @@
+TERMUX_SUBPKG_INCLUDE="bin/luantiserver bin/minetestserver"
+TERMUX_SUBPKG_DESCRIPTION="Headless server for the Luanti voxel game engine."
+TERMUX_SUBPKG_DEPENDS="jsoncpp, libandroid-spawn, libc++, libcurl, libgmp, libiconv, libpng, libsqlite, luajit, luanti-common, zstd"
```
Comment (truboxl) [inline x11-packages/luanti/luanti-server.subpackage.sh:3]:
```suggestion
TERMUX_SUBPKG_DEPENDS="jsoncpp, libandroid-spawn, libc++, libcurl, libgmp, libiconv, libluajit, libsqlite, luanti-common, zlib, zstd"
```

File: x11-packages/luanti/build.sh (hunk 3 above)
Comment (robertkirkman) [reply x11-packages/luanti/build.sh]: Thank you for the suggestion, I agree that for some of these it is difficult to tell what the correct dependencies to mark are because multiple have the same or similar effect (`libxi` depends on `libx11` so pulls it in, `zstd` depends on `zlib` so pulls it in, etc.) so I will try to go with your suggestions. 

The reason I had written `libglvnd-dev` instead of `opengl` is because if I write `pkg install opengl` on Termux, the necessary file `/data/data/com.termux/files/usr/include/GLES2/gl2.h` does not get installed. For some reason, though, when `opengl` is placed in a dependency list for `./build-package.sh` (not `pkg install`), `./build-package.sh` automatically installs `libglvnd-dev`, but I don't understand how it is able to because `libglvnd-dev` is not in the `TERMUX_PKG_DEPENDS`  or `TERMUX_PKG_BUILD_DEPENDS` lines of `mesa` or `libglvnd`. 

Do you happen to know how that is able to work?

Comment (truboxl) [reply x11-packages/luanti/build.sh]: `libglvnd-dev` is a subpackage of `libglvnd`. The package and it's subpackages are downloaded and extracted when using `-I`. AFAIK `TERMUX_PKG_ANTI_BUILD_DEPENDS` can be used to block them from pulling in.

**Review by truboxl — APPROVED** [review]
I will help merge by the end of the day if no one objects

File: x11-packages/luanti/build.sh (hunk 3 above)
Comment (robertkirkman) [reply x11-packages/luanti/build.sh]: Thank you for explaining, that helped me to understand how it works and find the place where the dependency is included, I have found what you describe now, "`for subpkg in self.subpkgs`... `self.deps.add(subpkg.name)`" in line 155 of `buildorder.py`.

https://github.com/termux/termux-packages/blob/eee8df6a96933e6d4d6b4fb9b97ecde47490a477/scripts/buildorder.py#L155

Putting a print statement in there that just prints out the variables `self.name` and `subpkg.name` while `buildorder.py` runs shows a confirmation of this,

```
[buildorder.py]: subpackage dependency of packge libglvnd included: libglvnd-dev
```

This must be why `libglvnd-dev`  gets installed by `build-package.sh` when building packages that depend on the `opengl` metapackage or the `libglvnd` package directly. I just wanted to make a note of what is going on here that makes the `opengl` metapackage work as a dependency instead of `libglvnd-dev`, so that if this `buildorder.py` behavior ever changes in the future and affects the dependencies of `luanti`, there would be a record of why the `opengl` dependency was used at this time.

**Comment by robertkirkman** [comment]
luanti has been copy striked on Google Play today; Google Play Termux and/or this Termux might be imminently copy striked for distributing luanti.

**Comment by TomJo2000** [comment]
That seems unlikely to me.
The DMCA takedown against Luanti is fraudulent and has been disputed by upstream.
https://blog.luanti.org/2026/08/27/luanti-dmca-tracer-ai/
